---------------------
PatchSet 342 
Date: 2011/08/02 10:39:48
Author: deraadt
Branch: HEAD
Tag: OPENBSD_5_0_BASE 
Log:
After the MANIFEST-based symbolic link creation loop, there is a check
for a specific link to see if things worked out.  Add a check for the very
last file in the MANIFEST, as well, since we are trying to spot a very
odd bug where symbolic links are not being created.  Hopefully this will
help us diagnose it.
ok guenther millert

Members: 
	Configure:1.23->1.24 

Index: src/gnu/usr.bin/perl/Configure
diff -u src/gnu/usr.bin/perl/Configure:1.23 src/gnu/usr.bin/perl/Configure:1.24
--- Configure:1.23	Fri Sep 24 15:06:38 2010
+++ Configure	Tue Aug  2 17:39:48 2011
@@ -2756,6 +2756,10 @@
 				echo "Failed to create the symlinks (t/base/lex.t missing).  Aborting." >&4
 				exit 1
 			fi
+			if test ! -f x2p/walk.c; then
+				echo "Failed to create the symlinks (x2p/walk.c missing).  Aborting." >&4
+				exit 1
+			fi
 			cd UU
 			;;
 		*)	echo "(I cannot figure out how to do symbolic links, ignoring mksymlinks)." >&4
---------------------
PatchSet 346 
Date: 2012/01/09 09:15:22
Author: millert
Branch: HEAD
Tag: OPENBSD_5_1_BASE 
Log:
More efficient method of building the symlink tree that makes better
use of awk.  Slightly faster and works around an apparent namei or
buffer cache related bug on arm.  Requested and OK deraadt@

Members: 
	Configure:1.24->1.25 

Index: src/gnu/usr.bin/perl/Configure
diff -u src/gnu/usr.bin/perl/Configure:1.24 src/gnu/usr.bin/perl/Configure:1.25
--- Configure:1.24	Tue Aug  2 17:39:48 2011
+++ Configure	Mon Jan  9 16:15:22 2012
@@ -2724,33 +2724,22 @@
 	*)	case "$lns:$issymlink" in
 		*"ln"*" -s:"*"test -"?)
 			echo "Creating the symbolic links..." >&4
-			echo "(First creating the subdirectories...)" >&4
 			cd ..
-			awk '{print $1}' $src/MANIFEST | grep / | sed 's:/[^/]*$::' | sort -u | while true; do
-				read directory
-				test -z "$directory" && break
-				mkdir -p $directory
-			done
+			awk -v src="$src" '{
+				dir=$1;
+				if (!sub(/\/[^\/]*$/, "", dir)) { dir = "." }
+				mf[dir] = mf[dir]" "src"/"$1;
+			} END {
+				for (d in mf) {
+					if (d != ".") { system("mkdir -p "d) }
+					system("ln -sf "mf[d]" "d);
+				}
+			}' $src/MANIFEST
 			# Sanity check 1.
 			if test ! -d t/base; then
 				echo "Failed to create the subdirectories.  Aborting." >&4
 				exit 1
 			fi
-			echo "(Then creating the symlinks...)" >&4
-			awk '{print $1}' $src/MANIFEST | while true; do
-				read filename
-				test -z "$filename" && break
-				if test -f $filename; then
-					if $issymlink $filename; then
-						rm -f $filename
-					fi
-				fi
-				if test -f $filename; then
-					echo "$filename already exists, not symlinking."
-				else
-					ln -s $src/$filename $filename
-				fi
-			done
 			# Sanity check 2.
 			if test ! -f t/base/lex.t; then
 				echo "Failed to create the symlinks (t/base/lex.t missing).  Aborting." >&4
